### PR TITLE
Optimize ReifyQuotes#mayChange

### PR DESCRIPTION
**Pull request by smarter**

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala [hunk 1]
```diff
@@ -620,7 +620,8 @@ class ReifyQuotes extends MacroTransformWithImplicits with InfoTransformer {
     transform(tp)
   }
 
-  override protected def mayChange(sym: Symbol)(implicit ctx: Context): Boolean = sym.is(Macro)
+  override protected def mayChange(sym: Symbol)(implicit ctx: Context): Boolean =
+    ctx.compilationUnit.containsQuotesOrSplices && sym.isTerm && sym.is(Macro)
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala]: `ctx.compilationUnit.containsQuotesOrSplices` will always be true. We only run this phase if it is true
https://github.com/lampepfl/dotty/pull/4261/files#diff-7b199cbe62cc044a10a26459c0f0fa46R94

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala]: I was wrong, we do need this check.

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala]: We could also avoid accessing the symbol 
```scala
  override def transform(ref: SingleDenotation)(implicit ctx: Context): SingleDenotation =
    if (ctx.compilationUnit.containsQuotesOrSplices) super.transform(ref) else ref

  override protected def mayChange(sym: Symbol)(implicit ctx: Context): Boolean = sym.isTerm && sym.is(Macro)
```

Comment (smarter) [reply compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala]: Calling `Denotation#symbol` is just a field access, so I don't think that's worth it.

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/ReifyQuotes.scala]: Ok

**Pull request merged**